File: sdk-python/gatedhouse/_factory.py
# Copyright (c) 2026 12vectors.com
# SPDX-License-Identifier: MIT
# See the LICENSE file in the repository root for the full license text.

"""Static factory mirroring Java's ``GatedhouseFactory``."""

from __future__ import annotations

from . import _migrator, _schema_check
from ._config import GatedhouseConfig
from ._exceptions import GatedhouseInitializationError
from ._gatedhouse import DefaultGatedhouse, Gatedhouse
from ._token_verifier_config import TokenVerifierConfig
from ._token_verifier_gatedhouse import JustTokenVerifierGatedhouse


class GatedhouseFactory:

    def __init__(self) -> None:
        raise TypeError("GatedhouseFactory is not instantiable")

    @staticmethod
    def create(config: GatedhouseConfig) -> Gatedhouse:
        if config is None:
            raise TypeError("config must not be None")
        _schema_check.verify(config.database)

        gatedhouse = DefaultGatedhouse(config)
        try:
            config.group_source.start(gatedhouse)
        except Exception as e:
            # Best-effort cleanup before propagating.
            try:
                config.group_source.close()
            except Exception:
                pass
            raise GatedhouseInitializationError(
                f"GroupSource.start failed during Gatedhouse initialization: {e}"
            ) from e
        return gatedhouse

    @staticmethod
    def create_just_token_verifier(config: TokenVerifierConfig) -> Gatedhouse:
        """Creates a lightweight Gatedhouse instance that only supports
        token verification and requires no database backend."""
        if config is None:
            raise TypeError("config must not be None")
        return JustTokenVerifierGatedhouse(config)

    @staticmethod
    def migrate(config: GatedhouseConfig) -> None:
        if config is None:
            raise TypeError("config must not be None")
        _migrator.migrate(config.database)

File: sdk-rust/src/factory.rs
//! Static factory mirroring Java's `GatedhouseFactory`.

use std::sync::Arc;

use crate::config::GatedhouseConfig;
use crate::error::GatedhouseError;
use crate::gatedhouse::{DefaultGatedhouse, Gatedhouse};
use crate::just_token_verifier::JustTokenVerifierGatedhouse;
use crate::migrator;
use crate::schema_check;
use crate::token_verifier::TokenVerifierConfig;

pub struct GatedhouseFactory;

impl GatedhouseFactory {
    /// Verify the schema is current and construct a [`Gatedhouse`].
    /// Returns an `Arc<dyn Gatedhouse>` so the host can clone it across
    /// threads without thinking about lifetimes.
    pub fn create(config: GatedhouseConfig) -> Result<Arc<dyn Gatedhouse>, GatedhouseError> {
        schema_check::verify(config.database.as_ref())?;

        let group_source = config.group_source.clone();
        let gatedhouse = Arc::new(DefaultGatedhouse::new(config));
        // GroupSource may want to capture a reference to Gatedhouse; pass
        // the trait object so it has the public surface.
        group_source.start(gatedhouse.as_ref() as &dyn Gatedhouse);

        Ok(gatedhouse as Arc<dyn Gatedhouse>)
    }

    /// Creates a lightweight Gatedhouse instance that only supports token
    /// verification and requires no database backend.
    ///
    /// Every database-backed method on the returned handle panics with
    /// "Database operations not supported on token-verifier-only
    /// instance" — the analog of Java's `UnsupportedOperationException`.
    pub fn create_just_token_verifier(config: TokenVerifierConfig) -> Arc<dyn Gatedhouse> {
        Arc::new(JustTokenVerifierGatedhouse::new(&config))
    }

    /// Run any pending migrations against the configured database. Safe
    /// to invoke concurrently from multiple processes — the runner takes
    /// a Postgres advisory lock.
    pub fn migrate(config: &GatedhouseConfig) -> Result<(), GatedhouseError> {
        migrator::migrate(config.database.as_ref())
    }
}
